src/app/components/media/MediaTimeline.js

```diff
@@ -0,0 +1,469 @@
+/* eslint-disable max-len */
+/* eslint-disable no-sequences */
+/* eslint-disable no-param-reassign */
+/* eslint-disable no-return-assign */
+/* eslint-disable no-console */
+/* eslint-disable no-shadow */
+import React, { Component } from 'react';
+import { Timeline } from '@meedan/check-ui';
+
+import {
+  createClip, renameClip, destroyClip, retimeClip,
+  createCommentThread, destroyComment, createComment, updateComment,
+  createTag, renameTag, destroyTag, retimeTag,
+  createPlace, createPlaceInstance, retimePlace, destroyPlace, renamePlace, repositionPlace,
+} from './MediaTimelineUtils';
+
+const NOOP = () => {};
+
+class MediaTimeline extends Component {
+  static getDerivedStateFromProps({
+    media: {
+      id: mediaId, dbid,
+      tags: { edges: tags = [] },
+      clips: { edges: clips = [] },
+      geolocations: { edges: locations = [] },
+      comments,
+    },
+    duration, time,
+    currentUser: { name: first_name, id: userId, profile_image: profile_img_url },
+  }) {
+    const projecttags = [];
+    const entities = {};
+
+    // console.log({
+    //   mediaId, dbid, tags, clips, comments, locations,
+    // });
+
+    const commentThreads = comments.edges.filter(({ node: { dbid, parsed_fragment } }) => !!dbid && !!parsed_fragment && !!parsed_fragment.t).map(({
+      node,
+      node: {
+        id: thread_id, dbid, text, parsed_fragment,
+        annotator: {
+          id: annoId, name, profile_image,
+        },
+        comments,
+      },
+    }) => ({
+      id: thread_id,
+      dbid,
+      text,
+      start_seconds: parsed_fragment.t[0],
+      user: {
+        id: annoId,
+        first_name: name,
+        last_name: '',
+        profile_img_url: profile_image,
+      },
+      replies: comments && comments.edges ? comments.edges.map(({
+        node: {
+          id, created_at, text,
+          annotator: {
+            id: annoId, name, profile_image,
+          },
+        },
+      }) => ({
+        id,
+        thread_id,
+        created_at,
+        text,
+        start_seconds: 0,
+        user: {
+          id: annoId,
+          first_name: name,
+          last_name: '',
+          profile_img_url: profile_image,
+        },
+      })).sort((a, b) => a.created_at - b.created_at) : [],
+      node,
+    }));
+
+    clips.filter(({ node: { id } }) => !!id).forEach(({
+      node,
+      node: { id, parsed_fragment: { t: [start_seconds, end_seconds] }, data: { label: name } },
+    }) => {
+      if (!entities[`clip-${name}`]) {
+        entities[`clip-${name}`] = {
+          id: `clip-${name}`,
+          name,
+          type: 'clip',
+          project_clip: { id: `clip-${name}`, name },
+          instances: [],
+          node,
+        };
+      }
+
+      entities[`clip-${name}`].instances.push({
+        id,
+        start_seconds,
+        end_seconds,
+        url: new URL(`#t=${start_seconds},${end_seconds}&id=${encodeURIComponent(id)}`, document.location.href).href,
```

Comment:
in creating the permalink of a clip

Reply:
Where do we display the permalink?

The problem I'm trying to forestall is: if the user _clicks_ the permalink, does that break the browser's Back button?

Reply:
It might because this permalink should be actually supported by your router, but I don't see how that can be done, to have both `/` and '#' handled by the router

Reply:
Can you please point me to the file and function that reads this `url`? I searched the diff and didn't find it.

Reply:
MediaComponent that `const { t = '', id } = qs.parse(document.location.hash.substring(1));`

Reply:
Aaugh we're talking over each other. @Laurian Thanks for the response, but that isn't the answer I'm looking for.

How does this `url`, which starts out as a property on the JavaScript Object being constructed on this line of code, end up in any user's `document.location`? Is there a `<Link>` somewhere? A `<a>`? Is it written to GraphQL? The question I'm asking is: where is the code that literally reads the `url` property off this object we're creating here?

I'll completely (and comically) rephrase the question. I see the story of this `url`, constructed here in `MediaTimeline.js` like this:

1. We build this`url`
2. ???
3. (maybe more ???)
4. User clicks the URL
5. Profit!

What is step 2? This is the line of code I'm not seeing. I probably sound stupid just for asking. But I couldn't find it, and my brain is fried. Please help me find it?

Reply:
there is a copy permalink hover menu in clips 
![image](https://user-images.githubusercontent.com/21679/85790598-22051a00-b739-11ea-9605-26b5b6ffca82.png)

Reply:
A-ha! Fantastic. Thank you for your patience. This is not a problem.